ChildProject/pipelines/metricsFunctions.py

```diff
@@ -0,0 +1,301 @@
+ - Create a new function using the same arguments (i.e. annotations, duration, **kwargs)
+ - Define calculation of the metric with:
+     - annotations, which is a dataframe containing all the relevant annotated segments  to use. It contains the annotation content (https://childproject.readthedocs.io/en/latest/format.html#id10) joined with the annotation index info (https://childproject.readthedocs.io/en/latest/format.html#id11) as well as any column that was requested to be added to the results by the user using --child-cols or --rec-cols (eg --child-cols child_dob,languages will make columns 'child_dob' and 'languaes' available)
+     - duration which is the duration of audio annotated in milliseconds
+     - kwargs, whatever keyword parameter you chose to pass to the function (except 'name', 'callable', 'set' which can not be used). This will need to be given with the list of metrics when called
+ - Wrap you function with the 'metricFunction' decorator to make it callable by the pipeline, read metricFunction help for more info
+"""
+
+#error message in case of missing columns in annotations
+MISSING_COLUMNS = 'The given set <{}> does not have the required column <{}> for computing the {} metric'
+    
+def metricFunction(args: set, columns: set, emptyValue = 0, name : str = None):
+    """Decorator for all metrics functions to make them ready to be called by the pipeline.
+    
+    :param args: set of required keyword arguments for that function, raise ValueError if were not given
+    :type args: set
+    :param columns: set of required columns in the dataframe given, missing columns raise ValueError
+    :type columns: set
+    :param name: default name to use for the metric in the resulting dataframe. Every keyword argument found in the name will be replaced by its value (e.g. 'voc_speaker_ph' uses kwarg 'speaker' so if speaker = 'CHI', name will be 'voc_chi_ph'). if no name is given, the __name__ of the function is used
+    :type name: str
+    :param emptyValue: value to return when annotations are empty but the unit was annotated (e.g. 0 for counts like voc_speaker_ph , None for proportions like lp_n)
+    :return: new function to substitute the metric function
+    :rtype: Callable
+    """
+    def decorator(function):
+        @functools.wraps(function)
+        def new_func(annotations: pd.DataFrame, duration: int, **kwargs):
+            for arg in args:
+                if arg not in kwargs : raise ValueError('{} metric needs an argument <{}>'.format(function.__name__,arg))
+            metname = name
+            if not name : metname = function.__name__            
+            for arg in kwargs:
+                metname = re.sub(arg , str(kwargs[arg]).lower(),metname)
+            if annotations.shape[0]:
+                for column in columns:
+                    if column not in annotations.columns : raise ValueError(MISSING_COLUMNS.format(annotations['set'].iloc[0],column,'voc_speaker_ph'))
```

Comment:
Totally nit-picky but as a user I'd prefer to learn about all the missing columns at once, so something like:

```python
missing_columns = columns - set(annotations.columns)
if missing_columns:
    raise ValueError(MISSING_COLUMNS.format(annotations['set'].iloc[0], missing_columns, metname))
```

Reply:
:+1:

Reply:
f61066db